src/java/org/apache/cassandra/schema/PartitionDenylist.java

```diff
@@ -0,0 +1,406 @@
+                   + denylistEntry.tokens.headSet(endToken, PartitionPosition.Kind.MAX_BOUND.equals(range.right.kind())).size();
+        }
+        catch (final ExecutionException e)
+        {
+            logAccessFailure(tid, e);
+            return 0;
+        }
+    }
+
+    private DenylistEntry readDenylist(final TableId tid)
+    {
+        final TableMetadata tmd = Schema.instance.getTableMetadata(tid);
+        if (tmd == null)
+            return new DenylistEntry();
+
+        final String readDenylist = String.format("SELECT * FROM %s.%s WHERE ks_name='%s' AND cf_name='%s' LIMIT %d", SystemDistributedKeyspace.NAME, PARTITION_DENYLIST_TABLE,
+                                                  tmd.keyspace, tmd.name, DatabaseDescriptor.maxdenyDenylistKeysPerTable() + 1);
+
+        try
+        {
+            final UntypedResultSet results = process(readDenylist, DatabaseDescriptor.denylistConsistencyLevel());
+            if (results == null || results.isEmpty())
+                return new DenylistEntry();
+
+            if (results.size() > DatabaseDescriptor.maxdenyDenylistKeysPerTable())
+                logger.error("Partition denylist for {}/{} has exceeded the maximum allowable size ({}). Remaining keys were ignored.", tmd.keyspace, tmd.name, DatabaseDescriptor.maxdenyDenylistKeysPerTable());
+
+            final Set<ByteBuffer> keys = new HashSet<>();
+            final NavigableSet<Token> tokens = new TreeSet<>();
+            for (final UntypedResultSet.Row row : results)
+            {
+                final ByteBuffer key = row.getBlob("key");
+                keys.add(key);
+                tokens.add(StorageService.instance.getTokenMetadata().partitioner.getToken(key));
+            }
+            return new DenylistEntry(ImmutableSet.copyOf(keys), ImmutableSortedSet.copyOf(tokens));
+        }
+        catch (final RequestExecutionException e)
+        {
+            logger.error("Error reading partition_denylist table for {}/{}", tmd.keyspace, tmd.name, e);
+            return null;
+        }
+    }
+
+    private Map<TableId, DenylistEntry> readAll()
+    {
+        final String readDenylist = String.format("SELECT * FROM %s.%s LIMIT %d", SystemDistributedKeyspace.NAME, PARTITION_DENYLIST_TABLE, DatabaseDescriptor.maxdenyDenylistKeysTotal() + 1);
+        try
+        {
+            final UntypedResultSet results = process(readDenylist, DatabaseDescriptor.denylistConsistencyLevel());
+            if (results == null || results.isEmpty())
+                return new HashMap<>();
+
+            if (results.size() > DatabaseDescriptor.maxdenyDenylistKeysTotal())
+                logger.error("Partition denylist has exceeded the maximum allowable total size ({}). Remaining keys were ignored.", DatabaseDescriptor.maxdenyDenylistKeysTotal());
+
+            final Map<TableId, Pair<Set<ByteBuffer>, NavigableSet<Token>>> allDenylists = new HashMap<>();
+            for (final UntypedResultSet.Row row : results)
```

Comment:
`I'm concerned that now (if the limit is exceeded) different nodes can have different denylist keys cache. So it would be pretty hard to troubleshoot.` I may be misunderstanding, but as the cache is populated via a query on a table w/a quorum CL, we'd have to "upgrade" that to CL.ALL to ensure that the data in the table, and subsequently the cache, is identical on all nodes especially in the face of overflow.

ISTM that the right answer is, as a user, to not exceed the max allowable partitions in your deny list cache so we don't cross that threshold into having to reason about edge conditions on multi-node consistency and cache load timing. Hence the warning, though perhaps it could be more prescriptive?

Updated warning to be more prescriptive on the downside of riding this edge; that + adding configurability on cache size and refresh interval via .yaml and JMX should cover it.

Reply:
> I may be misunderstanding

My point is like that: let's assume we have many node and 1000 denylisted keys (and maximum is 1000)
```
1
...
200
# 201 is not denylisted
202
...
1001
```
Nodes cached this data.

Then one more key (201) is denylisted. And now we have 1001 denylisted records.

Then one node is restarted and it reads first 1000 records. Now one node assumes 201 is denylisted and 1001 is not denylisted whereas other nodes assume opposite.

This is a corner case, but I think it may lead to pretty weird behavior of the system and we ideally need to prevent having such issues.

-----------------------------------------------------------
> Hence the warning, though perhaps it could be more prescriptive?

I do not know what the general policy on such cases (whether we assume that operators should use the tool properly or the tool should not allow to be used in wrong way). I like the latter because it is more safe.

Reply:
> This is a corner case, but I think it may lead to pretty weird behavior of the system and we ideally need to prevent having such issues.
+1. I would suggest either having a hard limit (like now) but blocking writes if limit is already reached, or not having a hard limit but logging warning about increased number of denylisted partitions. 
Debugging a corner case like this where sometimes partitions are denylisted and sometimes they are not, can be pretty hard in production.